cirq/protocols/unitary_test.py

```diff
@@ -74,3 +101,33 @@ def _unitary_(self) -> np.ndarray:
     # Explicit function should override
     assert cirq.has_unitary(FullyImplemented(True))
     assert not cirq.has_unitary(FullyImplemented(False))
+
+    # Test if decomposed operations _has_unitary_
+    a = cirq.NamedQubit('a')
+    b = cirq.NamedQubit('b')
+
+    assert cirq.has_unitary(cirq.GateOperation(Decomposable(True), (a, b)))
+    assert not cirq.has_unitary(cirq.GateOperation(ReturnsNotImplemented(),
+                                                   (a, b)))
+    assert cirq.has_unitary(cirq.GateOperation(DecomposableRecursion(True),
+                                               (a, b)))
+    assert not cirq.has_unitary(cirq.GateOperation(Decomposable(False), (a, b)))
+    assert not cirq.has_unitary(cirq.GateOperation(DecomposableRecursion(False),
+                                                   (a, b)))
+
+    assert cirq.has_unitary(cirq.GateOperation(DecomposableCombination(), (a,)))
+    assert not cirq.has_unitary(cirq.GateOperation(DecomposableCombination(),
+                                                   (a, b)))
+
+    # Test if decomposed operations has _unitary_
+    np.testing.assert_allclose(cirq.unitary(cirq.GateOperation(
+        Decomposable(True), (a,))), m)
+    np.testing.assert_allclose(cirq.unitary(cirq.GateOperation(
+        DecomposableRecursion(True), (a,))), m)
+
+    np.testing.assert_allclose(cirq.unitary(cirq.GateOperation(
```

Comment:
Make sure to exercise the case where sub-operations apply to a subset of qubits, or to qubits in a different order.